src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -58,33 +58,46 @@ internal class SemanticTokensRangeHandler : IRequestHandler<LSP.SemanticTokensRa
         /// 
         /// Null when the client does not support sending refresh notifications.
         /// </summary>
-        private readonly AsyncBatchingWorkQueue? _semanticTokenRefreshQueue;
+        private readonly AsyncBatchingWorkQueue<Uri?>? _semanticTokenRefreshQueue;
 
         #endregion
 
         public SemanticTokensRangeHandler(
             IGlobalOptionService globalOptions,
             IAsynchronousOperationListenerProvider asynchronousOperationListenerProvider,
             LspWorkspaceRegistrationService lspWorkspaceRegistrationService,
+            LspWorkspaceManager lspWorkspaceManager,
             ILanguageServerNotificationManager notificationManager,
             ClientCapabilities clientCapabilities)
         {
             _globalOptions = globalOptions;
             _asyncListener = asynchronousOperationListenerProvider.GetListener(FeatureAttribute.Classification);
 
             _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
-
             _disposalTokenSource = new();
 
             if (clientCapabilities.Workspace?.SemanticTokens?.RefreshSupport is true)
             {
-                // Only send a refresh notification to the client every 0.5s (if needed) in order to avoid
+                // Only send a refresh notification to the client every 2s (if needed) in order to avoid
                 // sending too many notifications at once.  This ensures we batch up workspace notifications,
                 // but also means we send soon enough after a compilation-computation to not make the user wait
                 // an enormous amount of time.
-                _semanticTokenRefreshQueue = new AsyncBatchingWorkQueue(
-                    delay: TimeSpan.FromMilliseconds(500),
-                    processBatchAsync: c => notificationManager.SendNotificationAsync(Methods.WorkspaceSemanticTokensRefreshName, c),
+                _semanticTokenRefreshQueue = new AsyncBatchingWorkQueue<Uri?>(
+                    delay: TimeSpan.FromMilliseconds(2000),
+                    processBatchAsync: (documentUris, cancellationToken) =>
+                    {
+                        var trackedDocuments = lspWorkspaceManager.GetTrackedLspText();
+                        foreach (var documentUri in documentUris)
+                        {
+                            if (documentUri is null || !trackedDocuments.ContainsKey(documentUri))
```

Comment:
note: processBatchAsync will get called on a threadpool thread.  Is it safe to make this check here on any thread?

Reply:
Hmm, I _think_ calling from any thread should be fine, none of this work is thread specific and I can't foresee any problems with concurrency